integration-tests/src/test_loop/utils/transactions.rs

```diff
@@ -201,3 +206,92 @@ pub fn get_node_data<'a>(node_datas: &'a [TestData], account_id: &AccountId) ->
     }
     panic!("RPC client not found");
 }
+
+/// Run a transaction until completion and assert that the result is "success".
+/// Returns the transaction result.
+pub fn run_tx(
+    test_loop: &mut TestLoopV2,
+    tx: SignedTransaction,
+    node_datas: &[TestData],
+    maximum_duration: Duration,
+) -> Vec<u8> {
+    let tx_res = execute_tx(test_loop, tx, node_datas, maximum_duration).unwrap();
+    assert_matches!(tx_res.status, FinalExecutionStatus::SuccessValue(_));
+    match tx_res.status {
+        FinalExecutionStatus::SuccessValue(res) => res,
+        _ => unreachable!(),
+    }
+}
+
+/// Submit a transaction and wait for the execution result.
+/// For invalid transactions returns an error.
+/// For valid transactions returns the execution result (which could have an execution error inside, check it!).
+pub fn execute_tx(
+    test_loop: &mut TestLoopV2,
+    tx: SignedTransaction,
+    node_datas: &[TestData],
+    maximum_duration: Duration,
+) -> Result<FinalExecutionOutcomeView, InvalidTxError> {
+    // Last node is usually the rpc node
+    let rpc_node_id = node_datas.len().checked_sub(1).unwrap();
+
+    let tx_hash = tx.get_hash();
+
+    let process_result = Arc::new(Mutex::new(None));
+    let process_result_clone = process_result.clone();
+
+    node_datas[rpc_node_id].client_sender.clone().send(MessageWithCallback {
+        message: ProcessTxRequest { transaction: tx, is_forwarded: false, check_only: false },
+        callback: Box::new(move |process_res| {
+            *process_result_clone.lock().unwrap() = Some(process_res);
+        }),
+    });
+
+    test_loop.run_until(
+        |tl_data| {
+            let mut processing_done = false;
+            if let Some(processing_outcome) = &*process_result.lock().unwrap() {
+                match processing_outcome.as_ref().unwrap() {
+                    ProcessTxResponse::NoResponse
+                    | ProcessTxResponse::RequestRouted
+                    | ProcessTxResponse::ValidTx => processing_done = true,
+                    ProcessTxResponse::InvalidTx(_err) => {
```

Comment:
In our test suit, do we ever expect to have invalid transactions sent? Might be a good idea to just panic at the callback site on line 246 instead?

Reply:
In either case, we aren't returning the InvalidTxError from this function as suggested by the signature.

Reply:
In my test I send a transaction that is invalid - one that hits the transaction size limit.
The function does return an `InvalidTxError`, lower in the code. I couldn't do it directly from `run_until` because it returns a boolean.

Reply:
Aah right, I missed the return statement on line 284